lib/route/link/bonding.c

```diff
@@ -19,6 +19,108 @@
 #include <netlink/route/link/bonding.h>
 #include <netlink-private/route/link/api.h>
 
+#define BOND_HAS_MODE		(1 << 0)
+#define BOND_HAS_ACTIVE_SLAVE	(1 << 1)
+
+struct bond_info {
+	uint8_t bn_mode;
+	uint32_t ifindex;
+	uint32_t bn_mask;
+};
+
+static int bond_info_alloc(struct rtnl_link *link)
+{
+	struct bond_info *bn;
+
+	if (link->l_info)
+		memset(link->l_info, 0, sizeof(*bn));
+	else {
+		bn = calloc(1, sizeof(*bn));
+		if (!bn)
+			return -NLE_NOMEM;
+
+		link->l_info = bn;
+	}
+
+	return 0;
+}
+
+static void bond_info_free(struct rtnl_link *link)
+{
+	_nl_clear_free(&link->l_info);
+}
+
+static int bond_put_attrs(struct nl_msg *msg, struct rtnl_link *link)
+{
+	struct bond_info *bn = link->l_info;
+	struct nlattr *data;
+
+	data = nla_nest_start(msg, IFLA_INFO_DATA);
+	if (!data)
+		return -NLE_MSGSIZE;
+	if (bn->bn_mask & BOND_HAS_MODE)
+		NLA_PUT_U8(msg, IFLA_BOND_MODE, bn->bn_mode);
+
+	if (bn->bn_mask & BOND_HAS_ACTIVE_SLAVE)
+		NLA_PUT_U32(msg, IFLA_BOND_ACTIVE_SLAVE, bn->ifindex);
+
+	nla_nest_end(msg, data);
+
+nla_put_failure:
+
+	return 0;
```

Comment:
This can only be reached if the `NLA_PUT*()` macros fail, because there is not enough space in the netlink message buffer. In fact, it leaves the message broken, because we don't even call `nla_nest_cancel()`.

This means, we really need to return an error code here. When looking at `git grep -A5 nla_put_failure:`, I think this indicates that it should be
```c
nla_put_failure:
    nla_nest_cancel(msg, data);
    return -NLE_MSGSIZE;
```

Reply:
Thanks for the detailed explanation. Taken care now.